### PR TITLE
[iOS] allow background download when app is not running

### PR DESCRIPTION
**Pull request by mehayhe**
This PR allows downloading files in background when app is not running. Currently the plugin cancels all download tasks when the app is terminated. For info about background download on iOS, check out the doc [here](https://developer.apple.com/documentation/foundation/url_loading_system/downloading_files_in_the_background?language=objc). 

This PR also fixes the follow issues: 
https://github.com/fluttercommunity/flutter_downloader/issues/675
https://github.com/fluttercommunity/flutter_downloader/issues/663

Steps for testing background download:

1. Build and install the example app on an iOS simulator or device
2. Run app from Home screen (background download might not work properly if the app has been started from xcode with a debugger)
3. start downloading a file
4. tap the Exit button on the top nav to exit the app gracefully. Do not kill app using task manager otherwise the any download tasks will be cancelled by the OS
5. wait a short while
6. open the example app again from the home screen
7. the downloading file should be completed with a “Ready” status

video demo:

https://user-images.githubusercontent.com/829645/176210370-9105153b-c104-4216-b971-2bf83baedd36.mov

### CODE REVIEW

**Comment by mehayhe** [comment]
@bartekpacia please review this PR when you have a chance

**Review by bartekpacia — APPROVED** [review]
Thank you so much for this PR! ❤️

My experience with Objective-C is almost `null`, so I can't really review it. @Hackmodford, I saw you in [this issue](https://github.com/fluttercommunity/flutter_downloader/issues/639). If you have some time, it'd be great if you could review this PR as well 🙏

**Comment by bartekpacia** [comment]
@mehayhe I meant to _Request Changes_ :)

File: example/lib/download_list_item.dart [hunk 1]
```diff
@@ -41,6 +43,11 @@ class DownloadListItem extends StatelessWidget {
             constraints: const BoxConstraints(minHeight: 32, minWidth: 32),
             icon: const Icon(Icons.play_arrow, color: Colors.green),
           ),
+          IconButton(
+            onPressed: () => onCancel!(task),
```
[the remark this answers is not in the document]
Comment (mehayhe) [reply example/lib/download_list_item.dart]: done

File: example/lib/download_list_item.dart [hunk 2]
```diff
@@ -57,7 +64,18 @@ class DownloadListItem extends StatelessWidget {
...
       );
     } else if (task.status == DownloadTaskStatus.canceled) {
-      return const Text('Canceled', style: TextStyle(color: Colors.red));
+      return Row(
+        mainAxisSize: MainAxisSize.min,
+        mainAxisAlignment: MainAxisAlignment.end,
+        children: [
+          const Text('Canceled', style: TextStyle(color: Colors.red)),
+          IconButton(
+            onPressed: () => onActionTap!(task),
```
[the remark this answers is not in the document]
Comment (mehayhe) [reply example/lib/download_list_item.dart]: done

File: ios/Classes/FlutterDownloaderPlugin.m [hunk 3]
```diff
@@ -562,9 +593,21 @@ - (void)registerCallbackMethodCall:(FlutterMethodCall*)call result:(FlutterResul
     NSArray *arguments = call.arguments;
     _callbackHandle = [arguments[0] longLongValue];
     _step = [arguments[1] intValue];
+    if (initialized) [self unqueueStatusEvents];
     result([NSNull null]);
 }
 
+- (void) unqueueStatusEvents {
+    @synchronized (self) {
+        //unqueue all pending download status events.
```
[the remark this answers is not in the document]
Comment (mehayhe) [reply ios/Classes/FlutterDownloaderPlugin.m]: done

File: ios/Classes/FlutterDownloaderPlugin.m [hunk 4]
```diff
@@ -347,6 +369,7 @@ - (NSString*)shortenSavedDirPath:(NSString*)absolutePath {
     
     if (absolutePath) {
         NSString* documentDirPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject];
+        if ([absolutePath isEqualToString:documentDirPath]) return @"";
```
Comment (Hackmodford) [inline ios/Classes/FlutterDownloaderPlugin.m:372]: Add braces to make harder to break in the future.

File: ios/Classes/FlutterDownloaderPlugin.m [hunk 5]
```diff
@@ -509,6 +532,7 @@ - (NSDictionary*)loadTaskWithId:(NSString*)taskId
         if (records != nil && [records count] > 0) {
             NSArray *record = [records firstObject];
             NSDictionary *task = [self taskDictFromRecordArray:record];
+            if (task.count == 0) return nil; // checking if task is valid
```
Comment (Hackmodford) [inline ios/Classes/FlutterDownloaderPlugin.m:535]: Add braces to make this harder to break in the future

**Comment by mehayhe** [comment]
@bartekpacia it's not possible to keep the download tasks active when terminating the app from app switcher. According to Apple:

> When you terminate an app via the multitasking UI (force quit), the system interprets that as a strong indication that the app should do no more work in the background, and that includes NSURLSession background tasks.

Check out this post from apple about NSURLSession: https://developer.apple.com/forums/thread/14855

**Comment by Hackmodford** [comment]
@bartekpacia Did my review comments not make it through? It was merged without them being addressed.

**Comment by bartekpacia** [comment]
I don't see any comments of yours :(

**Comment by Hackmodford** [comment]
Maybe I did something wrong. I highlighted the lines and typed a comment, then clicked the start review button

**Comment by mehayhe** [comment]
@Hackmodford let me know the lines you're referring to and the comment. I can create another PR if needed

**Comment by Hackmodford** [comment]
@mehayhe

I have learned how code reviews work on Github. :)
You should be able to see it now.

**Pull request merged**